EPMLSTRCMW-285 Refactor RunController API to be under Projects

Nikolai-Isaev opened the pull request:
feat: Add ProjectId to RunController route
feat: Add getRunsByProject functionality

Nikolai-Isaev committed:
feat: Add ProjectId to RunController route

- Move all methods under projects/{projectId}/runs
- add ProjectId for all methods and use it

Nikolai-Isaev committed:
feat: Add getRunsByProject functionality

DruzhininPavel commented on controllers/src/main/scala/cromwell/pipeline/controller/RunController.scala:
```diff
@@ -3,60 +3,69 @@ package cromwell.pipeline.controller
     path(RunId) { runId =>
       entity(as[RunUpdateRequest]) { runUpdateRequest =>
-        onComplete(runService.updateRun(runId, runUpdateRequest, accessToken.userId)) {
+        onComplete(runService.updateRun(runId, runUpdateRequest, projectId, accessToken.userId)) {
```
I think you should use a different combination of arguments for the function. Can you move projectId to start

Nikolai-Isaev replied:
Thought that that placing looks well in the meaning of russian-doll-like concept "run -> project -> user". So if we place project as a parameter it should go between run and user.

DruzhininPavel replied:
Ок

DruzhininPavel commented on controllers/src/test/scala/cromwell/pipeline/controller/RunControllerTest.scala:
```diff
@@ -34,102 +34,144 @@ class RunControllerTest
+          status shouldBe StatusCodes.InternalServerError
+        }
+      }
+
```
Can you not leave unused lines, and then the next time they are formatted, they will fall into the new MR

Nikolai-Isaev replied:
Fixed, thanks!

DruzhininPavel commented on services/src/main/scala/cromwell/pipeline/service/RunService.scala:
```diff
@@ -9,13 +9,15 @@ import scala.concurrent.{ ExecutionContext, Future }
+  def addRun(runCreateRequest: RunCreateRequest, projectId: ProjectId, userId: UserId): Future[RunId]
 
-  def getRunByIdAndUser(runId: RunId, userId: UserId): Future[Option[Run]]
+  def getRunByIdAndUser(runId: RunId, projectId: ProjectId, userId: UserId): Future[Option[Run]]
```
looks like this method should call getRunByIdAndProjectIdAndUser :)))

Nikolai-Isaev replied:
How about `getRunByWhateverYouWant` ? :) Btw, why don't we call it just `getRun`?

DruzhininPavel replied:
Maybe it would be better.

DruzhininPavel commented on services/src/test/scala/cromwell/pipeline/service/RunServiceTest.scala:
```diff
@@ -17,83 +18,128 @@ class RunServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
   private val runService: RunService = RunService(runRepository, projectService)
 
   "RunService" when {
+
```
Another void string.

Nikolai-Isaev replied:
That one is ambiguous. Don't know what style is preferable: with or without blank line in that case. We have different versions in our tests

DruzhininPavel replied:
Ok, sorry.

DruzhininPavel reviewed (changes requested):
Looks good but I think yo shoold make some small changes

DruzhininPavel reviewed (approved):
Ok.

gominasian reviewed (approved):
LGTM

myazinn reviewed (changes requested):
1) You commit message `refactor` does not correspond to the changes in the commit. `Refactor` means internal changes with little to no impact on business logic / external users. However, this commit significantly changes runs API, how is this a refactoring?
2) Suppose I have two projects: Project `Project _A` and Project `Project _B`. Let's assume each of them have one run: Run `Project_1_Run_1` for `Project_A` and Run `Project _B_Run_1` for `Project_B`. If I were to call `GET projects/ProjectA/runs/Project _B_Run_1`, I would expect to see 404 status code, because there is no `Project_B_Run_1` run for `ProjectA`. However, implementation in this PR will return `Project_B_Run_1` that belongs to `Project_B`. You must fix it.

Nikolai-Isaev commented:
>     1. You commit message `refactor` does not correspond to the changes in the commit. `Refactor` means internal changes with little to no impact on business logic / external users. However, this commit significantly changes runs API, how is this a refactoring?
> 
>     2. Suppose I have two projects: Project `Project _A` and Project `Project _B`. Let's assume each of them have one run: Run `Project_1_Run_1` for `Project_A` and Run `Project _B_Run_1` for `Project_B`. If I were to call `GET projects/ProjectA/runs/Project _B_Run_1`, I would expect to see 404 status code, because there is no `Project_B_Run_1` run for `ProjectA`. However, implementation in this PR will return `Project_B_Run_1` that belongs to `Project_B`. You must fix it.

1. Changed commit message type to "feat"
2. Added `projectId` checks to `RunService` get, delete, update methods. Added some corresponding testcases to `RunServiceTest`.

myazinn commented on services/src/main/scala/cromwell/pipeline/service/RunService.scala:
```diff
@@ -26,42 +28,55 @@ object RunService {
+      def getRunByIdAndUser(runId: RunId, projectId: ProjectId, userId: UserId): Future[Option[Run]] =
+        projectService.getUserProjectById(projectId, userId).flatMap { _ =>
+          runRepository.getRunByIdAndUser(runId, userId).flatMap {
+            case Some(run) if run.projectId == projectId => Future.successful(Some(run))
```
`Option#filter` ?

Nikolai-Isaev replied:
Fixed

myazinn commented on services/src/main/scala/cromwell/pipeline/service/RunService.scala:
```diff
@@ -26,42 +28,55 @@ object RunService {
-            runRepository.deleteRunById(runId)
-          case None => Future.failed(new RuntimeException("run with this id doesn't exist"))
+      def deleteRunById(runId: RunId, projectId: ProjectId, userId: UserId): Future[Int] =
+        projectService.getUserProjectById(projectId, userId).flatMap { _ =>
+          runRepository.getRunByIdAndUser(runId, userId).flatMap {
+            case Some(run) if run.projectId == projectId =>
+              runRepository.deleteRunById(runId)
+            case _ =>
+              Future.failed(new RuntimeException("run with this id doesn't exist"))
+          }
         }
```
DRY. Re-use `RunService#getRunByIdAndUser` here.

Nikolai-Isaev replied:
Done

myazinn commented on services/src/main/scala/cromwell/pipeline/service/RunService.scala:
```diff
@@ -26,42 +28,55 @@ object RunService {
-                cmwlWorkflowId = request.cmwlWorkflowId
+      def updateRun(runId: RunId, request: RunUpdateRequest, projectId: ProjectId, userId: UserId): Future[Int] =
+        projectService.getUserProjectById(projectId, userId).flatMap { _ =>
+          runRepository.getRunByIdAndUser(runId, userId).flatMap {
```
DRY. Re-use RunService#getRunByIdAndUser here.

Nikolai-Isaev replied:
Done

myazinn commented on services/src/test/scala/cromwell/pipeline/service/RunServiceTest.scala:
```diff
@@ -17,83 +18,161 @@ class RunServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
+
+      "fail with access denied exception" taggedAs Service in {
+        val error = ProjectAccessDeniedException()
+        when(runRepository.getRunByIdAndUser(runId, userId)).thenReturn(Future.failed(error))
```
This test makes no sense. Our repository is not supposed to return `ProjectAccessDeniedException`

Nikolai-Isaev replied:
Line 89 deleted

myazinn commented on services/src/test/scala/cromwell/pipeline/service/RunServiceTest.scala:
```diff
@@ -17,83 +18,161 @@ class RunServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
+
+      "fail with access denied exception" taggedAs Service in {
+        val error = ProjectAccessDeniedException()
+        when(runRepository.getRunByIdAndUser(runId, userId)).thenReturn(Future.failed(error))
```
You don't need this line, it is unreachable. And even if it was, our repository is not supposed to return failed future with such error.

Nikolai-Isaev replied:
That is definitely a strange one. Can't explain it other than by an alien invasion, thanks! Deleted

GrigoriiBerezin reviewed (approved):
Nicely done
